File: 05-08-2024_Bottom_View_of_a_Binary_Tree.py
class Solution:
    def bottomView(self,root):
        if not root:
            return []
    
        queue = deque([root])
        hd_node_map = {}
        node_hd_map = {root: 0}
    
        while queue:
            temp = queue.popleft()
            hd = node_hd_map[temp]
    
            hd_node_map[hd] = temp.data
    
            if temp.left:
                queue.append(temp.left)
                node_hd_map[temp.left] = hd - 1
            if temp.right:
                queue.append(temp.right)
                node_hd_map[temp.right] = hd + 1
    
        result = [hd_node_map[hd] for hd in sorted(hd_node_map)]
        return result

File: 05-08-2024_Bottom_View_of_a_Binary_Tree.java
class Solution
{
    //Function to return a list containing the bottom view of the given tree.
    public ArrayList<Integer> bottomView(Node root) {
        ArrayList<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }

        Map<Integer, Integer> hdNodeMap = new TreeMap<>();
        Map<Node, Integer> nodeHdMap = new HashMap<>();
        Queue<Node> queue = new LinkedList<>();

        queue.add(root);
        nodeHdMap.put(root, 0);

        while (!queue.isEmpty()) {
            Node temp = queue.poll();
            int hd = nodeHdMap.get(temp);
            
            hdNodeMap.put(hd, temp.data);

            if (temp.left != null) {
                queue.add(temp.left);
                nodeHdMap.put(temp.left, hd - 1);
            }
            if (temp.right != null) {
                queue.add(temp.right);
                nodeHdMap.put(temp.right, hd + 1);
            }
        }

        for (Map.Entry<Integer, Integer> entry : hdNodeMap.entrySet()) {
            result.add(entry.getValue());
        }

        return result;
    }
}
